cwms-data-api/src/main/java/cwms/cda/data/dao/JooqDao.java

```diff
@@ -434,6 +434,47 @@ private static InvalidItemException buildInvalidItem(RuntimeException input) {
         return exception;
     }
 
+    public static boolean isInvalidUnits(RuntimeException input) {
+        boolean retVal = false;
+
+        Optional<SQLException> optional = getSqlException(input);
+        if (optional.isPresent()) {
+            SQLException sqlException = optional.get();
+            String message = sqlException.getLocalizedMessage();
+            int errorCode = sqlException.getErrorCode();
+
+            retVal = errorCode == 20998
+                    && message.contains("ORA-20102: The unit")
+                    && message.contains("is not a recognized CWMS Database unit for the")
+                ;
+        }
+        return retVal;
+    }
+
+    private static InvalidItemException buildInvalidUnits(RuntimeException input) {
+
+        Throwable cause = input;
+        if (input instanceof DataAccessException) {
+            DataAccessException dae = (DataAccessException) input;
+            cause = dae.getCause();
+        }
+
+        String localizedMessage = cause.getLocalizedMessage();
+        if (localizedMessage != null) {
+            // skip ahead in localizedMessage to "ORA-20102:"
+            int start = localizedMessage.indexOf("ORA-20102:");
+            if (start >= 0) {
+                localizedMessage = localizedMessage.substring(start);
+                String[] parts = localizedMessage.split("\n");
+                if (parts.length >= 1) {
+                    localizedMessage = parts[0];
+                }
+            }
+        }
+
+        return new InvalidItemException(localizedMessage, cause);
```

Comment:
Needs the owasp sanitizer used here. The unit itself is user input so it needs to be properly escaped if sent back to the user; which in this case I think it should.

Otherwise the unit `steal user cookie values and do bad things` could cause some problem.

Reply:
Thats a great point

Reply:
I added this